File: python/lesson 11/subsets.py
# https://leetcode.com/problems/subsets/

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.subsetsImpl(nums, 0, [], result)
        return result
        
    def subsetsImpl(self, nums: List[int], start: int, current: List[int], result: List[List[int]]):
        result.append(current.copy())
    
        for i in range(start, len(nums)):
            current.append(nums[i])
            self.subsetsImpl(nums, i + 1, current, result)
            current.pop()

File: csharp/lesson 11/Subsets.cs
// https://leetcode.com/problems/subsets/

public class Solution
{
    public IList<IList<int>> Subsets(int[] nums)
    {
        var result = new List<IList<int>>();
        Subsets(nums, 0, new LinkedList<int>(), result);
        return result;
    }
    
    private void Subsets(int[] nums, int start, LinkedList<int> current, IList<IList<int>> result)
    {
        result.Add(new List<int>(current));
        
        for (var i = start; i < nums.Length; i++)
        {
            current.AddLast(nums[i]);
            Subsets(nums, i + 1, current, result);
            current.RemoveLast();
        }
    }
}
